File: solutions/ruby/yacht/2/yacht.rb
class Yacht
  def initialize(dice, category)
    @dice = dice
    @category = category
  end
  
  def score
    if @category == 'yacht'
      @dice.uniq.length == 1 ? 50 : 0
    elsif @category == 'ones'
      @dice.count(1) * 1
    elsif @category == 'twos'
      @dice.count(2) * 2
    elsif @category == 'threes'
      @dice.count(3) * 3
    elsif @category == 'fours'
      @dice.count(4) * 4
    elsif @category == 'fives'
      @dice.count(5) * 5
    elsif @category == 'sixes'
      @dice.count(6) * 6
    elsif @category == 'four of a kind'
      groups = @dice.group_by(&:itself)
      target_value = groups.find { |_, v| v.length >= 4 }&.first
      target_value ? target_value * 4 : 0
    elsif @category == 'full house'
      return 0 unless @dice.uniq.length == 2
      counts = @dice.group_by(&:itself).values.map(&:count).sort
      counts == [2, 3] ? @dice.sum : 0
    elsif @category == 'little straight'
      @dice.sort == [1, 2, 3, 4, 5] ? 30 : 0
    elsif @category == 'big straight'
      @dice.sort == [2, 3, 4, 5, 6] ? 30 : 0
    elsif @category == 'choice'
      @dice.sum
    end
  end
end

File: solutions/python/yacht/2/yacht.py
YACHT = "yacht"
ONES = "ones"
TWOS = "twos"
THREES = "threes"
FOURS = "fours"
FIVES = "fives"
SIXES = "sixes"
FULL_HOUSE = "full house"
FOUR_OF_A_KIND = "four of a kind"
LITTLE_STRAIGHT = "little straight"
BIG_STRAIGHT = "big straight"
CHOICE = "choice"


def score(dice, category):
    if category == YACHT:
        return 50 if len(set(dice)) == 1 else 0

    if category == ONES:
        return dice.count(1) * 1
    if category == TWOS:
        return dice.count(2) * 2
    if category == THREES:
        return dice.count(3) * 3
    if category == FOURS:
        return dice.count(4) * 4
    if category == FIVES:
        return dice.count(5) * 5
    if category == SIXES:
        return dice.count(6) * 6

    if category == FULL_HOUSE:
        if len(set(dice)) == 2 and all(dice.count(d) in (2, 3) for d in set(dice)):
            return sum(dice)
        return 0

    if category == FOUR_OF_A_KIND:
        for d in set(dice):
            if dice.count(d) >= 4:
                return d * 4
        return 0

    if category == LITTLE_STRAIGHT:
        return 30 if sorted(dice) == [1, 2, 3, 4, 5] else 0

    if category == BIG_STRAIGHT:
        return 30 if sorted(dice) == [2, 3, 4, 5, 6] else 0

    if category == CHOICE:
        return sum(dice)

    return 0